lib/src/collect_lcov.dart

```diff
@@ -51,93 +53,103 @@ class LcovPart {
-  Future<CoverageResult<List<File>>> _getCoverageJson(
-      String testFile, Directory coverageDir) async {
-    var args = [
-      "--coverage_dir=${coverageDir.path}",
-      "--package-root=${packageRoot}",
-      testFile
-    ];
-    var result = await processSystem.runProcess(Platform.executable, args);
-    if (result.exitCode < 0) {
-      stderr.writeln('stdout:');
-      stderr.writeln(result.stdout);
-      stderr.writeln('stderr:');
-      stderr.writeln(result.stderr);
-      throw new ProcessException(Platform.executable, args,
-          'There was a critical error. Exit code: ${result.exitCode}',
-          result.exitCode);
+  /// Formats coverage hitmap json to an lcov string
+  Future<String> _formatCoverageJson(Map<String, dynamic> hitmap){
+    var resolver;
+    if(packageRoot != null){
+      resolver = new Resolver(packageRoot: packageRoot, sdkRoot: sdkRoot);
     }
-    var reportFiles =
-        await coverageDir.list(recursive: false, followLinks: false).toList();
-    return new CoverageResult<List<File>>(reportFiles, result);
+    else{
+      resolver = new Resolver(packagesPath: packagesPath, sdkRoot: sdkRoot);
+    }
+    var formatter = new LcovFormatter(resolver);
+    return formatter.format(hitmap);
   }
-}
 
-class CoverageResult<E> {
-  final E result;
-  final ProcessResult processResult;
+  /// Generates and returns a coverage json file
+  Future<List<Map<String, dynamic>>> _getCoverageJson(
+      String testFile) async {
 
-  CoverageResult(this.result, this.processResult);
+    bool terminated = false;
 
-  /// Prints `processResult.stdout`
-  ///
-  /// If `processResult.exitCode` is not zero, also prints the exit code and
-  /// `processResult.stderr`.
-  void printSummary() {
-    print(processResult.stdout);
-    if (processResult.exitCode != 0) {
-      print("Process exited with code ${processResult.exitCode}.");
-      print(processResult.stderr);
+    var dartArgs = [
+      "--pause-isolates-on-exit",
+      "--enable-vm-service"
+    ];
+    if(packageRoot != null){
+      dartArgs.add("--package-root=${packageRoot}");
+    }
+    else{
+      dartArgs.add("--packages=${packagesPath}");
+    }
+    dartArgs.add(testFile);
+
+    Process process = await processSystem.startProcess(Platform.executable, dartArgs);
+    process.exitCode.then((exitCode){
+      if (exitCode < 0 && !terminated) {
+        throw new ProcessException(Platform.executable, dartArgs,
+            'There was a critical error. Exit code: ${exitCode}',
+            exitCode);
+      }
+    });
+
+    Completer<Uri> hostCompleter = new Completer<Uri>();
+    process.stdout.transform(UTF8.decoder).listen((data) {
+      Uri uri = util.extractObservatoryUri(data);
+      if(uri != null){
+        hostCompleter.complete(uri);
+      }
+    });
+    Uri host = await hostCompleter.future;
+
+    try{
+      Map<String, dynamic> coverageResults = await collect(host, true, false, timeout: new Duration(seconds: 60));
```

Comment:
In my testing, severe syntax errors in the test file given lead to `collect`to never complete. The timeout takes care of this case.